ports/espressif/supervisor/internal_flash.c

```diff
@@ -32,47 +33,114 @@
 
 #include "extmod/vfs.h"
 #include "extmod/vfs_fat.h"
+
 #include "py/mphal.h"
 #include "py/obj.h"
 #include "py/runtime.h"
-#include "lib/oofatfs/ff.h"
 
-#include "components/spi_flash/include/esp_partition.h"
+#include "esp_ota_ops.h"
+#include "esp_partition.h"
 
+#include "supervisor/filesystem.h"
 #include "supervisor/flash.h"
 #include "supervisor/usb.h"
 
-STATIC const esp_partition_t *_partition;
+#define OP_READ     0
+#define OP_WRITE    1
 
 // TODO: Split the caching out of supervisor/shared/external_flash so we can use it.
 #define SECTOR_SIZE 4096
 STATIC uint8_t _cache[SECTOR_SIZE];
 STATIC uint32_t _cache_lba = 0xffffffff;
 
+#if CIRCUITPY_STORAGE_EXTEND
+#if FF_MAX_SS == FF_MIN_SS
+#define SECSIZE(fs) (FF_MIN_SS)
+#else
+#define SECSIZE(fs) ((fs)->ssize)
+#endif // FF_MAX_SS == FF_MIN_SS
+STATIC DWORD fatfs_bytes(void) {
+    FATFS *fatfs = filesystem_circuitpy();
+    return (fatfs->csize * SECSIZE(fatfs)) * (fatfs->n_fatent - 2);
+}
+STATIC bool storage_extended = true;
+STATIC const esp_partition_t *_partition[2];
+#else
+STATIC const esp_partition_t *_partition[1];
+#endif // CIRCUITPY_STORAGE_EXTEND
+
 void supervisor_flash_init(void) {
-    _partition = esp_partition_find_first(ESP_PARTITION_TYPE_DATA,
+    if (_partition[0] != NULL) {
+        return;
+    }
+    _partition[0] = esp_partition_find_first(ESP_PARTITION_TYPE_DATA,
         ESP_PARTITION_SUBTYPE_DATA_FAT,
         NULL);
+    #if CIRCUITPY_STORAGE_EXTEND
+    _partition[1] = esp_ota_get_next_update_partition(NULL);
+    #endif
 }
 
 uint32_t supervisor_flash_get_block_size(void) {
     return FILESYSTEM_BLOCK_SIZE;
 }
 
 uint32_t supervisor_flash_get_block_count(void) {
-    return _partition->size / FILESYSTEM_BLOCK_SIZE;
+    #if CIRCUITPY_STORAGE_EXTEND
+    return ((storage_extended) ? (_partition[0]->size + _partition[1]->size) : _partition[0]->size) / FILESYSTEM_BLOCK_SIZE;
+    #else
+    return _partition[0]->size / FILESYSTEM_BLOCK_SIZE;
+    #endif
 }
 
 void port_internal_flash_flush(void) {
+}
 
+STATIC void single_partition_rw(const esp_partition_t *partition, uint8_t *data,
+    const uint32_t offset, const uint32_t size_total, const bool op) {
+    if (op == OP_READ) {
+        esp_partition_read(partition, offset, data, size_total);
+    } else {
+        esp_partition_erase_range(partition, offset, size_total);
+        esp_partition_write(partition, offset, _cache, size_total);
+    }
+}
+
+#if CIRCUITPY_STORAGE_EXTEND
+STATIC void multi_partition_rw(uint8_t *data,
+    const uint32_t offset, const uint32_t size_total, const bool op) {
+    if (offset > _partition[0]->size) {
```

Comment:
A suggestion: the `#if CIRCUITPY_STORAGE_EXTENDS`s below I think could be eliminated if you had an else branch here that did something like `#define multi_partition_rw(data, offset, size_total, op) single_partition_rw(_partition[0], data, offset, size_total, op)`, or a function to do the same. However, if you don't think that's a net improvement in readability that's fine as well.